src/ch/ch_monitor.c

```diff
@@ -563,25 +564,341 @@ chMonitorBuildSocketCmd(virDomainObjPtr vm, const char *socket_path)
+
+        if ((obj = virJSONValueFromString(buf))) {
+            if (virCHMonitorProcessEvent(mon, obj) < 0) {
+                VIR_WARN("Failed to process the event!");
+                ret = -1;
+            }
+            virJSONValueFree(obj);
+        } else {
+            VIR_WARN("Invalid JSON from monitor");
+            ret = -1;
+        }
+
+        *end = tmp;
+        buf = end;
+    }
+
+    /*
+     * If the buffer still has incomplete data, lets
+     * push it to the beginning.
+     */
+    sz = buf - mon->buffer;
+    if (sz < mon->buf_fill_sz) {
+        mon->buf_offset = mon->buf_fill_sz - sz;
+        memmove(mon->buffer, buf, sz);
+    } else {
+        mon->buf_offset = 0;
+    }
+
+    return ret;
+}
+
+static int virCHMonitorReadProcessEvents(virCHMonitorPtr mon)
+{
+    size_t max_sz = CH_MONITOR_BUFFER_SZ - mon->buf_offset;
+    char *buf = mon->buffer + mon->buf_offset;
+    virDomainObjPtr vm = mon->vm;
+    bool incomplete = false;
+    int events = 0;
+    size_t sz = 0;
+
+    memset(buf, 0, max_sz);
+    do {
+        ssize_t ret;
+
+        ret = read(mon->monitor_fd, buf + sz, max_sz - sz);
+        if (ret == 0 || (ret < 0 && errno == EINTR)) {
+            g_usleep(G_USEC_PER_SEC);
+            continue;
+        } else if (ret < 0) {
+            /*
+             * We should never reach here. read(2) says possible errors
+             * are EINTR, EAGAIN, EBADF, EFAULT, EINVAL, EIO, EISDIR
+             * We handle EINTR gracefully. There is some serious issue
+             * if we encounter any of the other errors(either in our code
+             * or in the system). Better to bail out.
+             */
+            VIR_ERROR("Failed to read monitor events!: %s", strerror(errno));
+            abort();
+        }
+
+        sz += ret;
+        events = virCHMonitorValidateEventsJSON(mon->buffer,
+                        mon->buf_offset + sz, &incomplete);
+        VIR_DEBUG("Monitor event(%ld):\n%s", sz, mon->buffer);
+
+    } while (virDomainObjIsActive(vm) && (sz < max_sz) &&
+             (events == 0 || incomplete));
+
+    /*
+     * We process the events from the read buffer if
+     *    - There is atleast one event in the buffer
+     *    - No incomplete events in the buffer or
+     *    - The buffer is full and may have incomplete entries.
+     *
+     * If the buffer is full, virCHMonitorProcessEvents processes
+     * the completed events in the buffer and moves incomplete
+     * entries to the start of the buffer and next read from the pipe
+     * starts from the offset.
+     */
+    mon->buf_fill_sz = sz + mon->buf_offset;
+    return (events > 0) ? virCHMonitorProcessEvents(mon, events) : events;
+}
+
 static void virCHMonitorEventLoop(void *data)
 {
     virCHMonitorPtr mon = (virCHMonitorPtr)data;
     virDomainObjPtr vm = mon->vm;
 
     VIR_DEBUG("Monitor event loop thread starting");
 
+    if (!(mon->buffer = g_malloc_n(sizeof(char), CH_MONITOR_BUFFER_SZ))) {
```

Comment:
g_malloc_n should abort on failure so this check for return value shouldn't be needed.

Reply:
Ahh I didn't know that. Thanks for clarifying, will fix it